cirq-core/cirq/transformers/transformer_api.py

```diff
@@ -0,0 +1,234 @@
+# Copyright 2022 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Defines the API for circuit transformers in Cirq."""
+
+import abc
+import functools
+from typing import (
+    Callable,
+    Tuple,
+    TypeVar,
+    Hashable,
+    Optional,
+    cast,
+    Type,
+    overload,
+)
+import dataclasses
+from enum import Enum
+
+from cirq.circuits.circuit import CIRCUIT_TYPE
+from cirq.circuits import circuit, frozen_circuit
+
+
+class LogLevel(Enum):
+    """Different logging resolution options for `TransformerStatsLogger`."""
+
+    INFO = 1
+    VERBOSE = 2
+    WARNING = 3
+
+
+class TransformerStatsLoggerBase:
```

Comment:
Why do we need a Base class + Implementation class here. It seems like the registerinitial and registerfinal would be the same for all transformers and the only thing changing would the circuits we are registering and the names. Am I missing something ?

Reply:
The different implementations would enable different kinds of loggers. For example, 
- One implementation could be a logger which works well for jupyter notebooks / html environments and plots an interactive toolbar to scroll through the action of different stages.
- While another logger could be more suited for a terminal based environment which outputs nicely formatted text to stdout / file. 

Different loggers could also choose to extract different kinds of information from the initial and final circuits passed for each optimization stage (eg: delta of number of 2q gates, no. of microwave gates etc.)

Reply:
That makes sense. I think I'm sold on doing this, one thing that still worries me a little is the erosion of precision on the different logging levels as the number of classes grow. So what does INFO vs WARNING mean in terminal vs on jupyter vs someone defining it for themselves in their own subclass implementation. Why couldn't one logger class gather all this data and then we leave it to the user to pull out no. of microwave gates of circuit depth or whatever ?

Reply:
- There are two places where the precision levels will be used -- one at the time of logging inside a transformer and second at the time of viewing the logs by the user. The first part will be the same for all logger implementations while the second part will be different for each logger. So, to answer your question, the logs for INFO or WARNING received by each logger implementation would be the same. But how would these different precision-level logs get rendered on a terminal vs a jupyter notebook would vary based on the subclass implementation. 
- One logger could gather all the data -- i.e `(initial_circuit, final_circuit, <list of all intermediate logs>)` for each transformer stage and store it. We could then make abstract methods for `logger.plot` / `logger.show_logs` which the subclasses could implement and do whatever data processing is needed at a later stage. This approach has a drawback that it limits us from writing more efficient loggers which need not necessarily store all this information. 

We could also provide the default implementation in the base class which stores all the data and let users who care about efficiency override these methods to define their custom loggers. I don't have strong opinions either way, but I'd prefer to do the implementation in a separate PR to keep the focus on the API implementation here.

Reply:
> We could also provide the default implementation in the base class which stores all the data and let users who care about efficiency override these methods to define their custom loggers

Let's do that and see how far we get :)

Reply:
You want me to do that in the current PR itself? I can send a follow-up PR since the two changes are unrelated (the transformer API vs logger implementation).

Reply:
Renamed the base class to `TransformerStatsLogger`.